aptos-move/aptos-release-builder/src/simulate.rs

```diff
@@ -0,0 +1,601 @@
+        &MODULE_ID_CREATE_SIGNER,
+        |m| {
+            let func_def = find_function_def_by_name(m, &FUNC_NAME_CREATE_SIGNER)
+                .ok_or_else(|| anyhow!("failed to locate `fun {}`", &*FUNC_NAME_CREATE_SIGNER))?;
+            func_def.visibility = Visibility::Public;
+
+            Ok(())
+        },
+    )
+}
+
+const MAGIC_FAILED_NEXT_EXECUTION_HASH_CHECK: u64 = 0xDEADBEEF;
+
+/// Patches `aptos_framework::aptos_governance::resolve_multi_step_proposal` so that
+/// it returns the requested signer directly, skipping the governance process altogether.
+fn patch_resolve_multistep_proposal(
+    state_view: &SimulationStateView<impl StateView>,
+    deserializer_config: &DeserializerConfig,
+    forbid_next_execution_hash: bool,
+) -> Result<()> {
+    use Bytecode::*;
+
+    patch_module(
+        state_view,
+        deserializer_config,
+        &MODULE_ID_APTOS_GOVERNANCE,
+        |m| {
+            let create_signer_handle_idx = get_or_add_simple_function_handle(
+                m,
+                MODULE_ID_CREATE_SIGNER.address,
+                MODULE_ID_CREATE_SIGNER.name.clone(),
+                FUNC_NAME_CREATE_SIGNER.clone(),
+                vec![SignatureToken::Address],
+                vec![SignatureToken::Signer],
+            );
+            let sig_u8_idx = get_or_add_signature(m, vec![SignatureToken::U8]);
+
+            let func_def = find_function_def_by_name(m, &FUNC_NAME_RESOLVE_MULTI_STEP_PROPOSAL)
+                .ok_or_else(|| {
+                    anyhow!(
+                        "failed to locate `fun {}`",
+                        &*FUNC_NAME_RESOLVE_MULTI_STEP_PROPOSAL
+                    )
+                })?;
+            func_def.acquires_global_resources = vec![];
+            let code = func_def.code.as_mut().ok_or_else(|| {
+                anyhow!(
+                    "`fun {}` must have a Move-defined body",
+                    &*FUNC_NAME_RESOLVE_MULTI_STEP_PROPOSAL
+                )
+            })?;
+
+            code.code.clear();
+            if forbid_next_execution_hash {
+                // If it is needed to forbid a next execution hash, inject additional Move
+                // code at the beginning that aborts with a magic number if the vector
+                // representing the hash is not empty.
+                //
+                //     if (!vector::is_empty(&next_execution_hash)) {
+                //         abort MAGIC_FAILED_NEXT_EXECUTION_HASH_CHECK;
+                //     }
+                //
+                // The magic number can later be checked in Rust to determine if such violation
+                // has happened.
+                code.code.extend([
+                    ImmBorrowLoc(2),
+                    VecLen(sig_u8_idx),
+                    LdU64(0),
+                    Eq,
+                    BrTrue(7),
+                    LdU64(MAGIC_FAILED_NEXT_EXECUTION_HASH_CHECK),
+                    Abort,
+                ]);
+            }
```

Comment:
@runtian-zhou here I check if the next execution hash is empty, and if so, abort with a magic number

Reply:
amazing!